src/join.js

```diff
@@ -108,61 +108,66 @@ if (canEvaluate) {
     };
 }}
 
+function makeJoinArgumentsError() {
+    return new TypeError(PROMISE_JOIN_TYPE_ERROR);
+}
 Promise.join = function () {
     var last = arguments.length - 1;
-    var fn;
-    if (last > 0 && typeof arguments[last] === "function") {
-        fn = arguments[last];
-        if (!__BROWSER__) {
-            if (last <= GENERATED_CLASS_COUNT && canEvaluate) {
-                var ret = new Promise(INTERNAL);
-                ret._captureStackTrace();
-                var HolderClass = holderClasses[last - 1];
-                var holder = new HolderClass(fn);
-                var callbacks = thenCallbacks;
+    if (arguments.length <= 1) {
+        return Promise.reject(makeJoinArgumentsError());
+    }
+    if(typeof arguments[last] !== "function") {
+        return Promise.reject(makeJoinArgumentsError());   
+    }
 
-                for (var i = 0; i < last; ++i) {
-                    var maybePromise = tryConvertToPromise(arguments[i], ret);
-                    if (maybePromise instanceof Promise) {
-                        maybePromise = maybePromise._target();
-                        var bitField = maybePromise._bitField;
-                        USE(bitField);
-                        if (BIT_FIELD_CHECK(IS_PENDING_AND_WAITING_NEG)) {
-                            maybePromise._then(callbacks[i], reject,
-                                               undefined, ret, holder);
-                            promiseSetters[i](maybePromise, holder);
-                            holder.asyncNeeded = false;
-                        } else if (BIT_FIELD_CHECK(IS_FULFILLED)) {
-                            callbacks[i].call(ret,
-                                              maybePromise._value(), holder);
-                        } else if (BIT_FIELD_CHECK(IS_REJECTED)) {
-                            ret._reject(maybePromise._reason());
-                        } else {
-                            ret._cancel();
-                        }
-                    } else {
-                        callbacks[i].call(ret, maybePromise, holder);
-                    }
+    var fn = arguments[last];
```

Comment:
I just removed two levels of nesting because the case where a function is not passed was removed - it's actually the same code mostly - you can see it in a side by side diff.